substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrSymbolRepository.java

```diff
@@ -87,12 +88,36 @@ private long getSymbolId(String imageHeapString, boolean replaceDotWithSlash) {
 
     @Override
     public void write(JfrChunkWriter writer) throws IOException {
+        // TODO: Epoch-based constant pool writing should allow for
+        // writing outside of a safepoint. However symbols can be
+        // marked in use directly by native events so the symbol
+        // repository probably needs to be epoch based as well
         assert VMOperation.isInProgressAtSafepoint();
         writer.writeCompressedLong(JfrTypes.Symbol.getId());
         writer.writeCompressedLong(table.getSize());
 
-        // TODO: iterate the table and write the symbol data in the correct encoding. Also consider
-        // symbol.getReplaceDotWithSlash() when writing the data.
+        JfrSymbol[] entries = table.getTable();
+        for (int i = 0; i < entries.length; i++) {
+            JfrSymbol entry = entries[i];
+            if (entry.isNonNull()) {
+                while (entry.isNonNull()) {
+                    writeSymbol(writer, entry);
+                    entry = entry.getNext();
+                }
+            }
+        }
+    }
+
+    private void writeSymbol(JfrChunkWriter writer, JfrSymbol symbol) throws IOException {
+        writer.writeCompressedLong(symbol.getId());
+        byte utf8 = 3;
+        writer.writeByte(utf8);
+        writer.writeCompressedInt(symbol.getValue().length());
+        String value = symbol.getValue();
+        if (symbol.getReplaceDotWithSlash()) {
+            value = value.replaceAll("\\.", "\\");
```

Comment:
Please call `String.replace(char, char)` instead (it uses far simpler code). It might also be possible to combine the conversion to UTF8 with this  replace (at the moment, two temporary copies of the String will be allocated).

Reply:
I've tweaked this to use replace(char, char) but I didn't quite follow the second recommendation. Do you mean to implement replaceDotWithSlash that acts on a UTF8 byte array?

Reply:
Yes, something like that is probably fairly simple to implement.

Reply:
Okay. I took a stab with the latest commit.